src/lib.rs

```diff
@@ -894,6 +909,33 @@ fn into_fr_plus_1<I: IntoFr>(x: I) -> Fr {
     result
 }
 
+fn derivation_index_into_fr(v: &[u8]) -> Fr {
+    // use number of rounds as a salt to avoid
+    // any child hash giving the same sequence
+    index_and_rounds_into_fr(v, 0)
+}
+
+fn index_and_rounds_into_fr(v: &[u8], rounds: u8) -> Fr {
+    let mut vr = vec![0u8; v.len() + 1];
+    vr[..v.len()].clone_from_slice(v);
+    vr[v.len()] = rounds;
+    let h = sha3_256(&vr);
+    match fr_from_be_bytes(h) {
+        Ok(fr) => {
+            // if fr is 0 or 1 do another round of hashing
+            // x * 0 = 0 which is a constant
+            // x * 1 = x which gives the same key
+            // it's extremely unlikely to find hash(vr) == 0 or 1
+            // so we could probably go without this check
+            if fr == Fr::zero() || fr == Fr::one() {
+                return index_and_rounds_into_fr(&h, rounds + 1);
+            }
+            fr
+        }
+        Err(_) => index_and_rounds_into_fr(&h, rounds + 1),
```

Comment:
So if the bytes can't be deserialized into fr, we hash them anyway.     Are we certain this is ok?   perhaps better to return a Result?     At the least, I think this case needs a comment to illustrate why it is ok to do it.

Reply:
The error is due to hash being outside the range of 0 to Fr::MAX, which means we need to do another round of hashing.

I've used a try/catch style implementation instead of if/then. The alternative would look something like

```
while hash > Fr::MAX {
  round = round + 1
  hash = sha3(hash + round)
}
```

The reason I didn't do the comparison is because comparing bytes with Fr is more complex than trying and failing. We could hardcode `FR_MAX_AS_BYTES` for the comparison but I feel that's adding more than is needed. If that const was exposed in crate `bls12_381` I'd use it but to my knowledge it's not available.

I think you've got a point though - try/catch isn't as semantically clear as the comparison.

Reply:
If it's faster to do it this way, I'd say leave it try/catch, but just comment what's going on so it is clearer to casual observers like myself.

Reply:
There's a test case for it: https://github.com/maidsafe/blsttc/pull/23/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R1648-R1653

> // index with many repeated hashes (34 rounds for this index)

This will hit that error 34 times before finally getting a hash less than Fr::MAX

Reply:
Comment about this is added here: https://github.com/maidsafe/blsttc/pull/23/files#diff-4007187965c1fea4fd252a76f5f1007b2ea59f6d86e2f454edee547c5036be24R76-R81